### PR TITLE
Stop creating static thread pool to fix random hang in onnx_test_runner

### PR DESCRIPTION
**Pull request by RandySheriffH**
The change is to scope the lifetime of the thread pool to be within the lifetime of ort env.
Otherwise, threads inside the thread pool calling the logging function will throw with "Default logger missing ..." causing a hang. These is a to-do item following this PR - in thread main loop functions, properly addressing exceptions/errors to avoid joining of early-exit threads.

### CODE REVIEW

File: onnxruntime/test/onnx/testenv.cc [hunk 1]
```diff
@@ -12,18 +12,9 @@
...
-  std::call_once(default_pool_init, [&env] {
-    using namespace onnxruntime::concurrency;
+std::unique_ptr<OrtThreadPool> TestEnv::CreateThreadPool(onnxruntime::Env& env) {
     int core_num = env.GetNumPhysicalCpuCores();
-
-    onnxruntime::ThreadOptions t_opts;
-    default_pool = std::make_unique<ThreadPool>(&env, t_opts, ORT_TSTR("onnx_runner_tp"), core_num, false);
-  });
-  return default_pool.get();
+    return std::make_unique<OrtThreadPool>(&env, onnxruntime::ThreadOptions{}, ORT_TSTR("onnx_runner_tp"), core_num, false);
```
Comment (pranavsharma) [inline onnxruntime/test/onnx/testenv.cc:17]: I wonder why this was called within call_once before. Hopefully this is not getting called in a loop somewhere. Can you confirm this?
Moreover, we might want to rename this function to CreateThreadPool to reflect its true job.

Comment (RandySheriffH) [reply onnxruntime/test/onnx/testenv.cc:17]: The function already renamed to "CreateThreadPool".
Let me double check very of its occurrence to double confirm.

Comment (RandySheriffH) [reply onnxruntime/test/onnx/testenv.cc:17]: https://github.com/microsoft/onnxruntime/blob/a81faee41ef2344de448caecb0f42a34fdc9ead7/onnxruntime/test/providers/cpu/model_tests.cc#L745
Should be the only other place calling the function.

**Review by pranavsharma — APPROVED** [review]

**Review by snnn — COMMENTED** [review]
Now I got the point.  onnx_test_runner.exe has a functionality that tests if we can run multiple inferences in parallel. And when it is happening, the calling thread of "onnxruntime::InferenceSession::Run()" doesn't have to be a thread that was created and managed by ONNX Runtime. Initially onnx_test_runner.exe used Windows thread pool on Windows. It didn't use ORT's thread pool.  Then, in order to support Linux, I copied eigen's SimpleThreadPoolTempl(not the one we are using) to this test program. When I did it, I thought for a test program simple is the best.  Then at some point we replaced it with the eigen thread pool in onnxruntime_common.  It was ok. Until the thread pool starts to take dependency on OrtEnv.  I think, as a test program, onnx_test_runner.exe should not rely on ORT's internal code too much. A defendant cannot use themself as the witness to prove the innocence. There must be someone else to help provide the evidence or testimony. This is the reason why I keep saying we need to keep this test program as simple as possible.

**Comment by snnn** [comment]
> The change is to shorten the lifetime of the thread pool to be within the lifetime of ort env.

We need to test what if the calling threads live longer than OrtEnv. For example, if someone created a thread pool, then use all the threads to run Ort inference concurrently, we can't ask them to shutdown all the threads before closing OrtEnv because these threads might have other uses.

**Comment by RandySheriffH** [comment]
> Now I got the point. onnx_test_runner.exe has a functionality that tests if we can run multiple inferences in parallel. And when it is happening, the calling thread of "onnxruntime::InferenceSession::Run()" doesn't have to be a thread that was created and managed by ONNX Runtime. Initially onnx_test_runner.exe used Windows thread pool on Windows. It didn't use ORT's thread pool. Then, in order to support Linux, I copied eigen's SimpleThreadPoolTempl(not the one we are using) to this test program. When I did it, I thought for a test program simple is the best. Then at some point we replaced it with the eigen thread pool in onnxruntime_common. It was ok. Until the thread pool starts to take dependency on OrtEnv. I think, as a test program, onnx_test_runner.exe should not rely on ORT's internal code too much. A defendant cannot use themself as the witness to prove the innocence. There must be someone else to help provide the evidence or testimony. This is the reason why I keep saying we need to keep this test program as simple as possible.

right - onnxruntime_perf_test uses Eigen thread pool instead ort thread pool, that's why it is cool from the issue.

**Comment by snnn** [comment]
@RandySheriffH, you masked the problem. You didn't solve it. You restricted the life time of the threads who calls InferenceSession::Run to be within the lifetime of ORT env, but it is not practical. We can't ask all our users to do so. We can't ask them to shutdown all their child threads before closing ORT env.

**Comment by snnn** [comment]
We need to restore the code to its original state ASAP, otherwise more bugs will sneak in.

**Pull request merged**